runtime-modules/project-token/src/tests/bonding_curve.rs

```diff
@@ -0,0 +1,1079 @@
+            Balances::set_balance(
+                Origin::root(),
+                Token::module_bonding_curve_reserve_account(token_id),
+                Balance::zero(),
+                Balance::zero(),
+            )
+            .unwrap();
+
+            let result = UnbondFixture::default().execute_call();
+
+            assert_err!(result, Error::<Test>::InsufficientJoyBalance);
+        })
+}
+
+#[test]
+fn unbonding_ok_with_user_joy_balance_correctly_increased() {
+    let (_, user_account) = member!(2);
+    let (user_account_id, user_balance) = (member!(2).1, joy!(5_000_000));
+    build_default_test_externalities_with_balances(vec![(user_account_id, user_balance)])
+        .execute_with(|| {
+            IssueTokenFixture::default()
+                .with_empty_allocation()
+                .execute_call()
+                .unwrap();
+            ActivateAmmFixture::default()
+                .with_creator_reward(Permill::zero())
+                .execute_call()
+                .unwrap();
+            BondFixture::default().execute_call().unwrap();
+            let user_reserve_pre = Balances::usable_balance(user_account);
+
+            UnbondFixture::default().execute_call().unwrap();
+
+            let user_reserve_post = Balances::usable_balance(user_account);
+            let correctly_computed_joy_amount = 95;
+            assert_eq!(
+                user_reserve_post - user_reserve_pre,
+                correctly_computed_joy_amount
+            );
+        })
+}
+
+#[test]
+fn unbonding_ok_with_user_crt_amount_correctly_decreased() {
+    let token_id = token!(1);
+    let creator_reward = Permill::from_percent(10);
+
+    let ((user_member_id, user_account_id), user_balance) = (member!(2), joy!(5_000_000));
+    build_default_test_externalities_with_balances(vec![(user_account_id, user_balance)])
+        .execute_with(|| {
+            IssueTokenFixture::default().execute_call().unwrap();
+            ActivateAmmFixture::default()
+                .with_creator_reward(creator_reward)
+                .execute_call()
+                .unwrap();
+            BondFixture::default().execute_call().unwrap();
+            let user_crt_pre =
+                Token::account_info_by_token_and_member(token_id, user_member_id).amount;
+
+            UnbondFixture::default()
+                .with_amount(DEFAULT_UNBONDING_AMOUNT)
+                .with_member_id(user_member_id)
+                .with_sender(user_account_id)
+                .execute_call()
+                .unwrap();
+
+            let user_crt_post =
+                Token::account_info_by_token_and_member(token_id, user_member_id).amount;
+            assert_eq!(user_crt_pre - user_crt_post, DEFAULT_UNBONDING_AMOUNT);
+        })
+}
+
+#[test]
+fn unbonding_ok_with_event_deposited() {
+    let token_id = token!(1);
+    let ((user_id, user_account), user_balance) = (member!(2), joy!(5_000_000));
+    let price = joy!(95); // see last test
```

Comment:
Please add explainer comment

Reply:
I have added an helper function with the full explicit formula i.e `a/2 * ((supply + amount)^2 - supply^2) + b * amount`